### PR TITLE
Drop `special_type` when aggregating PKs and FKs

### LINKED ISSUE #11208: Wrong type is inferred for count distinct in subsequent filter
To repo:
Screenshot 2019-10-23 23 02 24
(would expect to have number predicates)

Other aggregations work as expected.

### PR DESCRIPTION
**Pull request by paulrosenzweig**
Resolves #11208

The issue occurred because we always kept the base field's special type through an aggregation. In @sbelak's example from the issue, Product ID's 'type/FK' was retained when aggregated as "count distinct". That doesn't make sense and led to confusing post aggregation filter options.

The fix here drops the special type when aggregating PKs and FKs.

Though that solves the problem in the issue, I don't think it's enough. PKs and FKs should probably never keep their type through aggregation. The possible exception to this is min/max since that produces a value in the original set. That would still be a weird use case IMO. 

Other types should probably drop their type only for some aggregations. For example, you can sum a currency without changing the type. However, if you count the values in a currency column, you no longer have a currency. 

Questions
* Where besides post-aggregation filter options does this matter?
* Is it worth making this code more nuanced? (e.g. counted currencies are not currencies)

### CODE REVIEW

**Review by tlrobinson — COMMENTED** [review]
I think you're right it should be a little more nuanced. Doesn't this apply to the "distinct" aggregation rather than PK/FK specifically?

**Comment by paulrosenzweig** [comment]
@tlrobinson I figure it should apply to PK/FK for any aggregation. Except maybe min/max, but I'd wait to hear a good use case there.

Count and count distinct both invalidate other types. Maybe that it? So, the logic:

> Drop special_type if the column is a key or the aggregation is a count?

File: frontend/src/metabase-lib/lib/Dimension.js [hunk 1]
```diff
@@ -821,6 +821,8 @@ export class ExpressionDimension extends Dimension {
   }
 }
 
+const UNAGGREGATED_SPECIAL_TYPES = new Set([TYPE.FK, TYPE.PK]);
```
Comment (kdoh) [inline frontend/src/metabase-lib/lib/Dimension.js]: Might be worth just a small comment here explaining why FK and PK are special in this case per the PR discussion.

File: frontend/src/metabase-lib/lib/Dimension.js [hunk 2]
```diff
@@ -841,8 +843,10 @@ export class AggregationDimension extends Dimension {
 
   column(extra = {}) {
     const aggregation = this.aggregation();
+    const { special_type, ...column } = super.column();
     return {
-      ...super.column(),
+      ...column,
+      ...(!UNAGGREGATED_SPECIAL_TYPES.has(special_type) && { special_type }),
```
Comment (kdoh) [inline frontend/src/metabase-lib/lib/Dimension.js]: A similar quick comment on why we look for the `UNAGGREGATED_SPECIAL_TYPES`  here would be helpful.

**Review by kdoh — APPROVED** [review]
> So, the logic is: drop special_type if the column is a key or the aggregation is a count?

Based on the original issue and the discussion here I think that makes sense to me. Code looks good, just left a couple comment based comments cause I'm a curmudgeon.

**Pull request merged**